ext/standard/uuid.c

```diff
@@ -0,0 +1,662 @@
+		if (throw) {
+			throw_uuid_parse_exception_invalid_len(input, input_len, position, limit - position + 1);
+		}
+		return FAILURE;
+	}
+
+	for (; position <= limit; ++position) {
+		const char chr = input[position];
+
+		/* First digit of the byte. */
+		if (digit % 2 == 0) {
+			if ('0' <= chr && chr <= '9') {
+				byte = chr - '0';
+			}
+			else if ('a' <= chr && chr <= 'f') {
+				byte = chr - 'a' + 10;
+			}
+			else if ('A' <= chr && chr <= 'F') {
+				byte = chr - 'A' + 10;
+			}
+			else if (chr == '-') {
+				continue;
+			}
+			else {
+				if (throw) {
+					throw_uuid_parse_exception_invalid_char(input, input_len, position);
+				}
+				return FAILURE;
+			}
+		}
+		/* Second digit of the byte. */
+		else {
+			/* Shift upper half. */
+			byte *= 16;
+
+			if ('0' <= chr && chr <= '9') {
+				byte += chr - '0';
+			}
+			else if ('a' <= chr && chr <= 'f') {
+				byte += chr - 'a' + 10;
+			}
+			else if ('A' <= chr && chr <= 'F') {
+				byte += chr - 'A' + 10;
+			}
+			else if (chr == '-') {
+				continue;
+			}
+			else {
+				if (throw) {
+					throw_uuid_parse_exception_invalid_char(input, input_len, position);
+				}
+				return FAILURE;
+			}
+
+			if (digit > UUID_HEX_LEN) {
+				if (throw) {
+					throw_uuid_parse_exception(
+						input,
+						input_len,
+						position,
+						"Expected no more than %u hexadecimal digits",
+						UUID_HEX_LEN
+					);
+				}
+				return FAILURE;
+			}
+
+			uuid->bytes[digit / 2] = byte;
+		}
+
+		++digit;
+	}
+
+	if (digit < UUID_HEX_LEN) {
+		if (throw) {
+			throw_uuid_parse_exception_invalid_len(input, input_len, position, digit);
+		}
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+static zend_always_inline php_uuid *get_uuid(/*const*/ zval *uuid_object)
+{
+	zval *bytes = zend_read_property(php_ce_UUID, uuid_object, UUID_BYTES_PROP, UUID_BYTES_PROP_LEN, 1, NULL);
+
+	if (Z_TYPE_P(bytes) != IS_STRING) {
+		zend_throw_error(
+			zend_ce_type_error,
+			"Expected %s::$%s value to be of type %s, but found %s",
+			ZSTR_VAL(php_ce_UUID->name),
+			UUID_BYTES_PROP,
+			zend_get_type_by_const(IS_STRING),
+			zend_zval_type_name(bytes)
+		);
+		return NULL;
+	}
+
+	if (Z_STRLEN_P(bytes) != PHP_UUID_LEN) {
+		zend_throw_error(
+			zend_ce_error,
+			"Expected %s::$%s value to be exactly %u bytes long, but found %u",
+			ZSTR_VAL(php_ce_UUID->name),
+			UUID_BYTES_PROP,
+			PHP_UUID_LEN,
+			Z_STRLEN_P(bytes)
+		);
+		return NULL;
+	}
+
+	return (php_uuid *) Z_STRVAL_P(bytes);
+}
+
+static zend_always_inline void new_uuid(zval *object, const php_uuid *uuid)
+{
+	object_init_ex(object, php_ce_UUID);
+	zend_update_property_stringl(php_ce_UUID, object, UUID_BYTES_PROP, UUID_BYTES_PROP_LEN, (const char *) uuid->bytes, PHP_UUID_LEN);
+}
+
+PHP_METHOD(UUID, __construct)
+{
+	/* NOOP */
+}
+ZEND_BEGIN_ARG_INFO(UUID___construct_args, NULL)
+ZEND_END_ARG_INFO()
+
+PHP_METHOD(UUID, fromBinary)
+{
+	zval *input = NULL;
+
+	if (zend_parse_parameters_throw(ZEND_NUM_ARGS(), "z", &input) == FAILURE) {
```

Comment:
Please use correct zpp types, even if there are already arginfo typehints. This should use either `s` or `S`.

Reply:
This also applies to all other methods.